File: android/app/src/main/java/com/marianabuilds/charito/UsageStatsPlugin.java
package com.marianabuilds.charito;

import android.app.AppOpsManager;
import android.app.usage.UsageStats;
import android.app.usage.UsageStatsManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.os.Process;
import android.provider.Settings;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@CapacitorPlugin(name = "UsageStats")
public class UsageStatsPlugin extends Plugin {

    /**
     * Returns whether the app has been granted the PACKAGE_USAGE_STATS permission.
     * This is a "special" permission — the user must enable it manually in
     * System Settings > Apps > Special app access > Usage access.
     */
    private boolean hasUsagePermission() {
        AppOpsManager appOps = (AppOpsManager) getContext().getSystemService(Context.APP_OPS_SERVICE);
        int mode = appOps.checkOpNoThrow(
            AppOpsManager.OPSTR_GET_USAGE_STATS,
            Process.myUid(),
            getContext().getPackageName()
        );
        return mode == AppOpsManager.MODE_ALLOWED;
    }

    /**
     * Resolves a human-readable app label from a package name.
     * Falls back to a cleaned-up version of the package name if the app is not found.
     */
    private String getAppLabel(PackageManager pm, String packageName) {
        try {
            ApplicationInfo info = pm.getApplicationInfo(packageName, 0);
            return pm.getApplicationLabel(info).toString();
        } catch (PackageManager.NameNotFoundException e) {
            // Strip common prefixes to produce a cleaner fallback label
            String label = packageName;
            if (label.startsWith("com.")) label = label.substring(4);
            else if (label.startsWith("org.")) label = label.substring(4);
            else if (label.startsWith("net.")) label = label.substring(4);
            int dot = label.indexOf('.');
            return dot > 0 ? label.substring(0, dot) : label;
        }
    }

    /**
     * Plugin method called from JavaScript.
     *
     * Options:
     *   - days (int, default 7): how many days back to query
     *
     * Returns:
     *   { stats: Array<{ packageName, appName, totalTimeMs, launchCount, lastUsed }> }
     *
     * Rejects with "PERMISSION_DENIED" if usage access has not been granted.
     */
    @PluginMethod
    public void getUsageStats(PluginCall call) {
        if (!hasUsagePermission()) {
            call.reject("PERMISSION_DENIED");
            return;
        }

        Integer daysArg = call.getInt("days", 7);
        int days = daysArg != null ? daysArg : 7;

        UsageStatsManager usm = (UsageStatsManager) getContext()
                .getSystemService(Context.USAGE_STATS_SERVICE);
        PackageManager pm = getContext().getPackageManager();

        // Build query window: midnight `days` ago → now
        Calendar calendar = Calendar.getInstance();
        long endTime = calendar.getTimeInMillis();
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        calendar.add(Calendar.DAY_OF_YEAR, -(days - 1));
        long startTime = calendar.getTimeInMillis();

        List<UsageStats> usageStatsList = usm.queryUsageStats(
                UsageStatsManager.INTERVAL_DAILY, startTime, endTime);

        // Aggregate by package name (INTERVAL_DAILY may return multiple entries per package)
        Map<String, long[]> aggregated = new HashMap<>();
        // Values: [totalTimeMs, launchCount, lastUsed]
        if (usageStatsList != null) {
            for (UsageStats stat : usageStatsList) {
                String pkg = stat.getPackageName();
                long[] agg = aggregated.get(pkg);
                if (agg == null) {
                    agg = new long[]{0L, 0L, 0L};
                    aggregated.put(pkg, agg);
                }
                agg[0] += stat.getTotalTimeInForeground();
                // launchCount left as 0 — getAppLaunchCount() requires SDK level
                // that may not be available in all build environments
                if (stat.getLastTimeUsed() > agg[2]) {
                    agg[2] = stat.getLastTimeUsed();
                }
            }
        }

        // Build result — filter out entries with zero foreground time
        JSArray resultArray = new JSArray();
        for (Map.Entry<String, long[]> entry : aggregated.entrySet()) {
            String pkg = entry.getKey();
            long[] agg = entry.getValue();
            if (agg[0] <= 0) continue;

            JSObject obj = new JSObject();
            obj.put("packageName", pkg);
            obj.put("appName", getAppLabel(pm, pkg));
            obj.put("totalTimeMs", agg[0]);
            obj.put("launchCount", (int) agg[1]);
            obj.put("lastUsed", new java.util.Date(agg[2]).toInstant().toString());
            resultArray.put(obj);
        }

        JSObject result = new JSObject();
        result.put("stats", resultArray);
        call.resolve(result);
    }

    /**
     * Opens the Android "Usage Access" system settings screen so the user can
     * manually grant the PACKAGE_USAGE_STATS permission.
     */
    @PluginMethod
    public void openUsageAccessSettings(PluginCall call) {
        Intent intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
        call.resolve();
    }
}

File: android/app/src/main/java/com/marianabuilds/charito/UsageStatsPlugin.kt
package com.marianabuilds.charito

import android.app.AppOpsManager
import android.app.usage.UsageStats
import android.app.usage.UsageStatsManager
import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.os.Process
import android.provider.Settings
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.Calendar

@CapacitorPlugin(name = "UsageStats")
class UsageStatsPlugin : Plugin() {

    /**
     * Returns whether the app has been granted the PACKAGE_USAGE_STATS permission.
     * This is a "special" permission that cannot be granted via a normal runtime request —
     * the user must manually enable it in System Settings > Apps > Special app access > Usage access.
     */
    private fun hasUsagePermission(): Boolean {
        val appOps = context.getSystemService(Context.APP_OPS_SERVICE) as AppOpsManager
        val mode = appOps.checkOpNoThrow(
            AppOpsManager.OPSTR_GET_USAGE_STATS,
            Process.myUid(),
            context.packageName
        )
        return mode == AppOpsManager.MODE_ALLOWED
    }

    /**
     * Try to resolve a human-readable app label from a package name.
     * Falls back to the package name itself if the app is not found.
     */
    private fun getAppLabel(pm: PackageManager, packageName: String): String {
        return try {
            val info: ApplicationInfo = pm.getApplicationInfo(packageName, 0)
            pm.getApplicationLabel(info).toString()
        } catch (e: PackageManager.NameNotFoundException) {
            // Strip common prefixes to produce a cleaner fallback label
            packageName
                .removePrefix("com.")
                .removePrefix("org.")
                .removePrefix("net.")
                .split(".")
                .firstOrNull() ?: packageName
        }
    }

    /**
     * Plugin method called from JavaScript.
     *
     * Options:
     *   - days (Int, default 7): how many days back to query
     *
     * Returns:
     *   { stats: Array<{ packageName, appName, totalTimeMs, launchCount, lastUsed }> }
     *
     * Error codes:
     *   - PERMISSION_DENIED: user has not granted Usage Access — JS side should show a prompt
     */
    @PluginMethod
    fun getUsageStats(call: PluginCall) {
        if (!hasUsagePermission()) {
            call.reject("PERMISSION_DENIED", "Usage Access permission not granted")
            return
        }

        val days = call.getInt("days", 7) ?: 7

        val usm = context.getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager
        val pm = context.packageManager

        // Query window: midnight `days` ago → now
        val calendar = Calendar.getInstance()
        val endTime = calendar.timeInMillis
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        calendar.add(Calendar.DAY_OF_YEAR, -(days - 1))
        val startTime = calendar.timeInMillis

        val usageStatsList: List<UsageStats> =
            usm.queryUsageStats(UsageStatsManager.INTERVAL_DAILY, startTime, endTime)
                ?: emptyList()

        // Aggregate by package (INTERVAL_DAILY may return multiple entries per package)
        data class Agg(
            var totalTimeMs: Long = 0L,
            var launchCount: Int = 0,
            var lastUsed: Long = 0L
        )
        val aggregated = mutableMapOf<String, Agg>()
        for (stat in usageStatsList) {
            val pkg = stat.packageName
            val agg = aggregated.getOrPut(pkg) { Agg() }
            agg.totalTimeMs += stat.totalTimeInForeground
            // launchCount is only available on API 28+
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
                agg.launchCount += stat.appLaunchCount
            }
            if (stat.lastTimeUsed > agg.lastUsed) agg.lastUsed = stat.lastTimeUsed
        }

        // Build the JSON result — filter out system/launcher noise (totalTimeMs == 0)
        val resultArray = JSArray()
        aggregated
            .filter { (_, agg) -> agg.totalTimeMs > 0 }
            .entries
            .sortedByDescending { it.value.totalTimeMs }
            .forEach { (pkg, agg) ->
                val obj = JSObject()
                obj.put("packageName", pkg)
                obj.put("appName", getAppLabel(pm, pkg))
                obj.put("totalTimeMs", agg.totalTimeMs)
                obj.put("launchCount", agg.launchCount)
                obj.put("lastUsed", java.util.Date(agg.lastUsed).toInstant().toString())
                resultArray.put(obj)
            }

        val result = JSObject()
        result.put("stats", resultArray)
        call.resolve(result)
    }

    /**
     * Sends the user to the system "Usage Access" settings screen so they can
     * grant the PACKAGE_USAGE_STATS permission manually.
     * Called from JS when getUsageStats returns PERMISSION_DENIED.
     */
    @PluginMethod
    fun openUsageAccessSettings(call: PluginCall) {
        val intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        context.startActivity(intent)
        call.resolve()
    }
}
